Add Backend Validation for Note

iivic opened the pull request:
SOL-2204
The only requirement is that note must be type str or unicode (basestring).
I changed the ValidationError handling in the update method so notice that as well.
@vkaracic @marjev  @mjfrey

edx-webhook commented:
Thanks for the pull request, @iivic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=1140&repo=edx%2Fecommerce"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

vkaracic commented on ecommerce/extensions/catalogue/models.py:
```diff
@@ -4,13 +4,28 @@
+        self.clean()
+        super(Product, self).save(*args, **kwargs)  # pylint: disable=bad-super-call
+
+    def clean(self):
```
``Product`` already has a [clean](https://github.com/django-oscar/django-oscar/blob/1.1.1/src/oscar/apps/catalogue/abstract_models.py#L367) method. This is how it works for ``note`` attribute (or any [text](https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/catalogue/migrations/0016_coupon_note_attribute.py#L18) type attributes):
```
p = Product.objects.filter(product_class__name='Coupon').first()
p  # <Product: Test1>
p.attr.note = 123
p.clean()
ValidationError: [u"note attribute [u'Must be str or unicode']"]
```

Try to either re-use the already existing validation, or adjust it to your needs but don't override ``clean`` like this.

marjev commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -910,6 +910,32 @@ def test_update_coupon_with_invalid_max_uses(self, max_uses):
+        })
+        self.client.post(COUPONS_LINK, json.dumps(self.data), 'application/json')
+        coupon = Product.objects.get(title=self.data['title'])
+        details = self.get_response_json('GET', reverse('api:v2:coupons-detail', args=[coupon.id]))
```
Lines 927 and 928 are called consecutively in almost all test methods for updating coupon fields. It would be a good time to refactor this.

marjev commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -312,7 +312,12 @@ def update(self, request, *args, **kwargs):
+                msg=error.message
+            )
+            logger.exception(error_message)
+            raise serializers.ValidationError(error_message)
```
why not use `log_message_and_raise_validation_error `?

iivic replied:
This is not the same [ValidationError](http://www.django-rest-framework.org/api-guide/exceptions/#validationerror) as the django ValidationError.

marjev commented on ecommerce/extensions/catalogue/models.py:
```diff
@@ -4,13 +4,28 @@
+        try:
+            if not isinstance(self.attr.note, basestring) and self.attr.note is not None:
+                log_message_and_raise_validation_error(
+                    'Failed to create Product. Product note value must be type string'
```
nitpick: `must be of type string`.

marjev commented on ecommerce/extensions/catalogue/tests/test_models.py:
```diff
@@ -0,0 +1,39 @@
+import ddt
+
```
remove this space.

marjev commented on ecommerce/extensions/catalogue/tests/test_models.py:
```diff
@@ -0,0 +1,39 @@
+            id=1,
+            title='Test title',
+            product_class=product_class
+        )
```
Don't we have factories for `Product_class` and `Product`? If not, it's time to create those or update existing if, for any reason, you're not able to use those to create test objects.

iivic commented:
Ready for second review.

vkaracic commented on ecommerce/extensions/catalogue/models.py:
```diff
@@ -1,16 +1,28 @@
+                self.clean()
+            except ValidationError:
+                log_message_and_raise_validation_error(
+                    'Failed to create Product. Product note value must be of type string.'
```
``clean()`` does a validation of all the product attributes (among other properties, [see table](https://github.com/django-oscar/django-oscar/blob/1.1.1/src/oscar/apps/catalogue/abstract_models.py#L367)). If one of the other than ``note`` fails you will log this.
Lose the attribute check and just do this:
```
try:
    self.clean()
except ValidationError as e:
    log_message_and_raise_validation_error(e.message)
```

vkaracic commented on ecommerce/extensions/catalogue/tests/test_models.py:
```diff
@@ -0,0 +1,42 @@
+            name='Note',
+            product_class=coupon_product_class,
+            type='text'
+        )
```
You could re-use this: https://github.com/edx/ecommerce/blob/master/ecommerce/coupons/tests/mixins.py#L146

vkaracic commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -944,13 +943,37 @@ def test_update_coupon_with_invalid_max_uses(self, max_uses):
+            'title': 'Coupon update with note {}'.format(invalid_note),
+        })
+        self.client.post(COUPONS_LINK, json.dumps(self.data), 'application/json')
+        details = self._get_coupon_details(self.data['title'])
```
Nit: the POST call above is above every call of this abstraction, maybe factor it in as well?

iivic commented:
Ready for another review.

vkaracic commented on ecommerce/extensions/catalogue/models.py:
```diff
@@ -1,16 +1,28 @@
     history = HistoricalRecords()
 
+    def save(self, *args, **kwargs):
+        if hasattr(self.attr, 'note'):
```
Why would you call ``clean`` only for coupons with ``note``? Shouldn't everything be validated?

iivic replied:
Because the validation will fail if the note isn't set for a coupon product. I'm validating the field from the coupon creation/edit form. If you want to make the validation for everything I wish you good luck.

vkaracic replied:
Because it isn't set? That's odd since it's set to not be required: https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/catalogue/migrations/0016_coupon_note_attribute.py#L19
What's the error message that you get?

iivic replied:
It is not required when not calling clean() on Product but when clean is called it will require all the attributes to be set (title, coupon_vouchers, note). The title and coupon_vouchers will be set on every coupon but the note field can be empty and that's why I first check if the attribute is set.
The message is "note attribute cannot be blank"

vkaracic replied:
I found why that is happening and a couple of ways to fix it, BUT I'm not telling :smiley: 
It's a fun problem to tinker about (at least it was fun for me) and since ``note`` validation is not of crucial importance right now, I'd like to see what you'll come up with. Hopefully by the end of this exercise you'll be a bit more familiar by the inner workings of Django and Oscar.
To be clear, the assignment is to find a way to call ``clean()`` in the ``save()`` method successfully.
Have fun! :)
If necessary, I can give hints in exchange for free candy.

vkaracic commented on ecommerce/extensions/catalogue/tests/test_models.py:
```diff
@@ -0,0 +1,42 @@
+            title=self.COUPON_PRODUCT_TITLE,
+            product_class=self.coupon_product_class
+        )
+        voucher = factories.VoucherFactory(code='TESTVOUCHER')
```
Do you need the ``code`` parameter?

iivic replied:
No copy/paste leftovers.

iivic commented:
Ditto on the coupon factory. It will make testing easier.

vkaracic commented on ecommerce/extensions/catalogue/utils.py:
```diff
@@ -68,11 +68,17 @@ def create_coupon_product(
+    coupon_product.save()
+    coupon_product.attr.coupon_vouchers.coupon = coupon_product
+    coupon_product.attr.coupon_vouchers.save()
+
```
I would instantiate CouponVouchers separately:
```
coupon_vouchers = CouponVouchers()
...
coupon_vouchers.save()
...
vouchers = create_vouchers(...)
coupon_vouchers.add(*vouchers)
```

iivic commented:
I removed the oscar clean method call before save because a lot of tests need to be rewritten (around 520 tests failed). Now just the note type s being tested. I also moved the coupon_vouchers from create_vouchers method since it makes no sense keeping it there.

iivic commented:
@marjev  @vkaracic  Please take a quick look before I merge this.

vkaracic commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -357,6 +357,12 @@ def _get_voucher_range_with_updated_dynamic_catalog_values(self):
+        """Helper method that creates and returns coupon details."""
+        self.client.post(COUPONS_LINK, json.dumps(self.data), 'application/json')
+        coupon = Product.objects.get(title=self.data['title'])
+        return self.get_response_json('GET', reverse('api:v2:coupons-detail', args=[coupon.id]))
```
Since this returns only details, is it possible to create the coupon directly without sending a POST request to the coupons endpoint and use that to get the details? That will save up some testing time.

iivic replied:
I guess a Product could be created but we want to test if the api endpoint works as expected because of the validation so not calling the api would make the tests useless. If you have a better solution feel free to advice me.

vkaracic commented on ecommerce/extensions/catalogue/tests/test_models.py:
```diff
@@ -0,0 +1,42 @@
+    def test_create_product_with_note(self):
+        """Verify creating a product with valid note value creates product."""
+        note = 'Some test note.'
+        self._create_coupon_product_with_note_attribute(note)
```
This method should return the coupon product back, so you can use that one instead of the making another db call below.

iivic commented:
@vkaracic Comment addressed the other needs further explanation.

marjev commented on ecommerce/extensions/catalogue/models.py:
```diff
@@ -1,9 +1,10 @@
-# noinspection PyUnresolvedReferences
```
Why is this being removed?

marjev commented on ecommerce/extensions/catalogue/models.py:
```diff
@@ -13,6 +14,16 @@ class Product(AbstractProduct):
+                    'Failed to create Product. Product note value must be of type string.'
+                )
+        except AttributeError:
+            pass
```
Instead of overriding `save`, can you use this:
https://github.com/django-oscar/django-oscar/blob/31ea8dcbcce4ccb6bcbdbb08c3fe9b01103fae2c/src/oscar/apps/catalogue/product_attributes.py#L36

If not, you can use:
https://docs.djangoproject.com/en/1.10/ref/signals/#pre-save

iivic replied:
`validate_attributes` will make a bunch of tests fail (around 560) so I decided to validate the field directly instead of fixing every test.

Overriding the save method is consistent throughout the code. If you think pre-save is the way to go we should use it in the future as well.

marjev commented on ecommerce/extensions/catalogue/models.py:
```diff
@@ -30,6 +41,4 @@ def __unicode__(self):
         )
 
-
-# noinspection PyUnresolvedReferences
```
Again, why exactly is this being removed?

marjev commented on ecommerce/extensions/voucher/utils.py:
```diff
@@ -346,7 +346,7 @@ def _generate_code_string(length):
     return voucher_code
 
 
-def _create_new_voucher(code, coupon, end_datetime, name, offer, start_datetime, voucher_type):
```
Why remove `coupon`?

If you need to remove it, however, change the docstring.

vkaracic commented on ecommerce/extensions/catalogue/tests/test_models.py:
```diff
@@ -1,10 +1,36 @@
+import ddt
 from django.utils.timezone import now, timedelta
+from django.core.exceptions import ValidationError
```
Alphabetize please.
